src/plugins/reference/README.md

```diff
@@ -0,0 +1,128 @@
+  might be a mistake. The same is true, if `.` is used anywhere, but at the start of the 
+  reference.
+
+### Construction of the reference graph
+
+The process starts with a key `X`, which has the metakey `check/reference` set to the value
+`recursive`. This key `X` must be an array key. The basename of `X` will be called the `refname`.
+For each element in the array run the following process:
+
+1. Interpret the current key `K`s value as a referemce and throw an error, if the reference
+   is not valid.
+2. If there is no node for `K` in the reference graph create one. If there is no node for 
+   `R` in the reference graph create one. Insert an edge from `K` to `R` into the graph.
+3. Find a key `K1` directly below `R` that has the `refname` as its basename. Interpret this 
+   key as an array, if found. For each array element, start the process recursively from 1 
+   and then continue with 4.
+4. Find any key `K2` directly below `R` that has the metakey `check/reference` set to 
+   `alternative`. Interpret any such key as an array. For each array element, start the 
+   process recursively from 1, but this time using the basename of `K2` (the of the array)
+   as the `refname`.
+
+Once the whole process is finished, we will have a graph of the whole reference structure 
+stemming from `X`. This graph can then be check for acyclicity.
+
+### Restriction of references
+
+Without any additional intervention, the plugin accepts the name of any existing key as valid
+reference value. Existing in this context means, the key either has a value or metadata 
+associated with it. In recursive mode a key is also said to exist, if a key with the current 
+`refname` as a basename exists (has value or metadata) directly below the key in question. 
+If, however, the key containing the reference value has the `check/reference/restrict` metakey 
+set (even if the value is empty), its value has to be taken into account.
+
+If the value `check/reference/restrict` is anything but a valid array-element-name, i.e. a `#` 
+followed by `n` underscores (`_`) and `n+1` digits (`0-9`), its value will be used directly. 
+Otherwise `check/reference/restrict` has to be a valid array and its elements will be treated 
+as alternative restriction. Only one of these restriction has to be fulfilled for a reference to
+be valid.
+
+The restrictions themselves are UNIX-like pathname globbing expression, as defined by 
+[`fnmatch(3)`](http://man7.org/linux/man-pages/man3/fnmatch.3.html) using the `FNM_PATHNAME` 
+option. Path segments equal to `.` and `..` are interpreted like they are in UNIX-paths. If 
+the restriction starts with a `/` it will be treated as an absolute path, and the use of `.` 
+and `..` is unnecessary and discouraged. Therefore a warning will be emitted, if `.` or `..` 
+are encountered. The same is true if a path segment other then the first one is `.`.
+
+If a keyname matches the globbing expression it will be considered a valid reference 
```

Comment:
See the spec plugin, Elektra has small differences in globbing characters.

_ matches key names (except arrays) and # matches array entries. See src/libs/elektra/keyname.c

Reply:
Good question, it was rather by accident that * still works but it somehow makes sense to be not incompatible to glob(3) as people already know it (but only extend it to only match (non-)arrays).

And yes, a good idea to move the matching code. Unfortunately "fnmatch" might cause the library to be removed and there are already quite many libs depending on libease. So maybe it is better to create a libglob?